rust/agama-lib/src/storage/model.rs

```diff
@@ -504,14 +504,45 @@ pub struct DeviceInfo {
 pub struct BlockDevice {
     pub active: bool,
     pub encrypted: bool,
-    pub recoverable_size: DeviceSize,
     pub size: DeviceSize,
+    pub shrinking: ShrinkingInfo,
     pub start: u64,
     pub systems: Vec<String>,
     pub udev_ids: Vec<String>,
     pub udev_paths: Vec<String>,
 }
 
+#[derive(Debug, Clone, Serialize, Deserialize, utoipa::ToSchema)]
+#[serde(rename_all = "camelCase")]
+pub enum ShrinkingInfo {
+    Supported(DeviceSize),
+    Unsupported(Vec<String>),
+}
+
+impl TryFrom<zbus::zvariant::Value<'_>> for ShrinkingInfo {
+    type Error = zbus::zvariant::Error;
+
+    fn try_from(value: zbus::zvariant::Value) -> Result<Self, zbus::zvariant::Error> {
+        let hash: HashMap<String, OwnedValue> = value.clone().try_into()?;
+        let mut info: Option<Self> = None;
+
+        if let Some(size) = get_optional_property(&hash, "Supported")? {
+            info = Some(Self::Supported(size));
+        }
+        if let Some(reasons) = get_optional_property(&hash, "Unsupported")? {
+            info = Some(Self::Unsupported(reasons));
+        }
+
+        if let Some(info_value) = info {
```

Comment:
What about using [ok_or_else](https://doc.rust-lang.org/std/option/enum.Option.html#method.ok_or_else) or [ok_or](https://doc.rust-lang.org/std/option/enum.Option.html#method.ok_or)?

```rust
info.ok_or_else(|| Self::Error::Message(format!("Wrong value for Shrinking: {}", value)))

```

(`ok_or` is shorter, but it builds a string that may be unnecessary).

Additionally, the call to `to_string` is not needed (the `format!` macro already returns a string).

Reply:
Good point! Thanks.